## Given Task(A)
### HandsOn - Predict The Data Scientists Salary In India from kaggle use sklearn
dataset: Predict-The-Data-Scientists-Salary-In-India_Train_Dataset.csv

    Data Features set:

    1.Name of the company (Encoded)
    2.Years of experience(split to min and max experience),
    3.Job description,
    4.Job designation,
    5.Job Type,
    6.Key skills,
    7.Location (needs LabelEncoded),
    8.Salary in Rupees Lakhs(To be predicted)(needs LabelEncoded).
    

Find the best estimator among below :
    
    clf1 = DecisionTreeClassifier(),
    clf2 = RandomForestClassifier(n_estimators=100),
    clf3 = ExtraTreesClassifier(n_estimators=100),
    clf4 =  AdaBoostClassifier(n_estimators=100),
    clf5 =  GradientBoostingClassifier(n_estimators=100).

In [4]:
from sklearn.pipeline import * 
from sklearn.naive_bayes import * 
from sklearn.cluster import *  
from sklearn.covariance import *  
from sklearn.cross_decomposition import *  
from sklearn.datasets import *  
from sklearn.decomposition import *  
from sklearn.ensemble import *  
from sklearn.feature_extraction import *  
from sklearn.feature_extraction.text import *  
from sklearn.feature_selection import *  
from sklearn.gaussian_process import *  
from sklearn.linear_model import *  
from sklearn.manifold import *  
from sklearn.metrics import *  
from sklearn.mixture import *  
from sklearn.model_selection import *  
from sklearn.neighbors import *  
from sklearn.neural_network import *  
from sklearn.preprocessing import *  
# from sklearn_pandas import DataFrameMapper
from sklearn.svm import *  
from sklearn.tree import *
from yellowbrick.classifier import ConfusionMatrix, ROCAUC

In [5]:
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt

In [71]:
path = r"code\data\Predict-The-Data-Scientists-Salary-In-India_Train_Dataset.csv"
data = pd.read_csv(path)
data.head(10)

Unnamed: 0 experience                                    job_description  \
0           0    5-7 yrs  Exp: Minimum 5 years;Good understanding of IOC...   
1           1  10-17 yrs  He should have handled a team of atleast 5-6 d...   
2           2    5-9 yrs  Must be an effective communicator (written & s...   
3           3   7-10 yrs  7  -  10 years of overall experience in data e...   
4           4    1-3 yrs  Chartered Accountancy degree or MBA in Finance...   
5           5   5-10 yrs  4.\tWillingness to travel;5.\tExperience in Te...   
6           6  13-15 yrs  - Experience in managing fraud analytics with ...   
7           7   6-10 yrs  Qualified Chartered Accountant with 1-3 yrs of...   
8           8    1-3 yrs  Proven experience as payroll executive or spec...   
9           9    2-6 yrs  Other activities as may be required over the p...   

                                           job_desig   job_type  \
0        Senior Exploit and Vulnerability Researcher        NaN   
1                                           Head SCM        NaN   
2  Deputy Manager - Talent Management & Leadershi...  Analytics   
3                 Associate Manager Data Engineering  Analytics   
4                            TS- GSA- Senior Analyst        NaN   
5                          Senior Trainer- Academics        NaN   
6  Associate Vice President - Fraud Analytics - S...  Analytics   
7               Manager - Internal Audit - CA - Big4        NaN   
8                                  Payroll Executive        NaN   
9  Required CA (manager Finance) for R&R Salons P...        NaN   

                                          key_skills               location  \
0  team skills, communication skills, analytical ...  Delhi NCR(Vikas Puri)   
1  ppc, logistics, inventory management, supply c...                Sonepat   
2  HR Analytics, Employee Engagement, Training, S...              Delhi NCR   
3  SQL, Javascript, Automation, Python, Ruby, Ana...              Bengaluru   
4  accounting, finance, cash flow, financial plan...                Gurgaon   
5  academics, effective communication, teacher tr...                  Noida   
6  Fraud Analytics, People Management Skills, Tea...              Delhi NCR   
7  Accounting, Internal Audit, Auditing, Enterpri...                   Pune   
8  esic, hr, bonus, pf, tds, payroll, superannuat...    Ahmedabad(Bodakdev)   
9  accounting standards, finance, auditing, finan...   Bengaluru(Kodihalli)   

   salary  company_name_encoded  
0   6to10                  3687  
1  10to15                   458  
2  15to25                  4195  
3  10to15                   313  
4    3to6                  1305  
5   6to10                  4428  
6  25to50                  1988  
7   6to10                  1793  
8    0to3                  3398  
9  15to25                  3444

In [72]:
data = data.drop(columns= ['Unnamed: 0',])
data.columns

Index(['experience', 'job_description', 'job_desig', 'job_type', 'key_skills',
       'location', 'salary', 'company_name_encoded'],
      dtype='object')

In [73]:
data['job_type'].value_counts()

job_type
Analytics    3747
analytics     921
ANALYTICS      77
analytic       32
Analytic       20
Name: count, dtype: int64

In [74]:
data.isnull().sum()

experience                  0
job_description          4418
job_desig                   0
job_type                15005
key_skills                  1
location                    0
salary                      0
company_name_encoded        0
dtype: int64

In [75]:
data.shape

(19802, 8)

In [76]:
data['MinExp'] = [int(x.split('-')[0].strip('yrs')) for x in data['experience']]
data['MaxExp'] = [int(x.split('-')[1].strip('yrs')) for x in data['experience']]
data.columns

Index(['experience', 'job_description', 'job_desig', 'job_type', 'key_skills',
       'location', 'salary', 'company_name_encoded', 'MinExp', 'MaxExp'],
      dtype='object')

In [77]:

data['key_skills'] = data['key_skills'].fillna(data['key_skills'].mode()[0])

In [78]:
data['experience']= data['experience'].str.replace('yrs','')
data['experience'].loc[0:10]

0       5-7 
1     10-17 
2       5-9 
3      7-10 
4       1-3 
5      5-10 
6     13-15 
7      6-10 
8       1-3 
9       2-6 
10      3-7 
Name: experience, dtype: object

In [79]:
data['salary'].unique()

array(['6to10', '10to15', '15to25', '3to6', '25to50', '0to3'],
      dtype=object)

In [85]:
#Preprocessing - LableEncoder
le= LabelEncoder()
data['location'] = le.fit_transform(data['location'])
data['salary'] = le.fit_transform(data['salary'])
data['job_desig']= le.fit_transform(data['job_desig'])
data['job_type']= le.fit_transform(data['job_type'])
data.head()

job_desig  job_type  location  salary  company_name_encoded  MinExp  MaxExp
0       9104         5       597       5                  3687       5       7
1       4395         5      1412       1                   458      10      17
2       3058         2       525       2                  4195       5       9
3       1087         2       114       1                   313       7      10
4      10385         5       811       4                  1305       1       3

In [81]:
data = data.drop(columns=['job_description','experience'])
data.head()

job_desig   job_type  \
0        Senior Exploit and Vulnerability Researcher        NaN   
1                                           Head SCM        NaN   
2  Deputy Manager - Talent Management & Leadershi...  Analytics   
3                 Associate Manager Data Engineering  Analytics   
4                            TS- GSA- Senior Analyst        NaN   

                                          key_skills  location  salary  \
0  team skills, communication skills, analytical ...       597       5   
1  ppc, logistics, inventory management, supply c...      1412       1   
2  HR Analytics, Employee Engagement, Training, S...       525       2   
3  SQL, Javascript, Automation, Python, Ruby, Ana...       114       1   
4  accounting, finance, cash flow, financial plan...       811       4   

   company_name_encoded  MinExp  MaxExp  
0                  3687       5       7  
1                   458      10      17  
2                  4195       5       9  
3                   313       7      10  
4                  1305       1       3

In [82]:
data = data.drop(columns=['key_skills'])
data.head()

job_desig   job_type  location  \
0        Senior Exploit and Vulnerability Researcher        NaN       597   
1                                           Head SCM        NaN      1412   
2  Deputy Manager - Talent Management & Leadershi...  Analytics       525   
3                 Associate Manager Data Engineering  Analytics       114   
4                            TS- GSA- Senior Analyst        NaN       811   

   salary  company_name_encoded  MinExp  MaxExp  
0       5                  3687       5       7  
1       1                   458      10      17  
2       2                  4195       5       9  
3       1                   313       7      10  
4       4                  1305       1       3

In [86]:
data.head()

job_desig  job_type  location  salary  company_name_encoded  MinExp  MaxExp
0       9104         5       597       5                  3687       5       7
1       4395         5      1412       1                   458      10      17
2       3058         2       525       2                  4195       5       9
3       1087         2       114       1                   313       7      10
4      10385         5       811       4                  1305       1       3

In [87]:
X = data.drop(['salary'], axis=1)
y = data['salary']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

In [89]:
models = [
    ("Decision Tree", DecisionTreeClassifier()),
    ("Random Forest", RandomForestClassifier(n_estimators=100)),
    ("Extra Trees", ExtraTreesClassifier(n_estimators=100)),
    ("AdaBoost", AdaBoostClassifier(n_estimators=100)),
    ("Gradient Boosting", GradientBoostingClassifier(n_estimators=100))
]

for name, model in models:
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    print(f"{name} Accuracy: {accuracy_score(y_test, y_pred):.4f}")

Decision Tree Accuracy: 0.3433
Random Forest Accuracy: 0.3716
Extra Trees Accuracy: 0.3598
AdaBoost Accuracy: 0.3961
Gradient Boosting Accuracy: 0.4499


## Hence, GradientBoost is the best model based on accuracy (0.4499).